models/trie_mimc/frontend_proof.go

```diff
@@ -0,0 +1,486 @@
+	}
+	return lsb
+}
+
+// Right shift by N Bytes
+func rightShiftNBytes(api frontend.API, input frontend.Variable, N int) frontend.Variable {
+	var lsb frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	if N >= 32 {
+		return 0
+	}
+	for i := 0; i < N*8; i++ {
+		lsb = api.Add(lsb, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+
+	var divider frontend.Variable = 1
+	for i := 0; i < N; i++ {
+		divider = api.Mul(divider, 256)
+	}
+	return api.DivUnchecked(api.Sub(input, lsb), divider)
+}
+
+func hashVectors(api frontend.API, hFunc mimc.MiMC,
+	hashes ...frontend.Variable) frontend.Variable {
+
+	hFunc.Write(hashes[0])
+	hFunc.Write(rightShift1Byte(api, hashes[1]))
+	for n, h := range hashes[2:] {
+		hFunc.Write(api.Add(NBytesLeftShift(api, hashes[n+1], n+1),
+			rightShiftNBytes(api, h, n+2)))
+	}
+	hFunc.Write(leftShift1Byte(api, leastNBytes(api, hashes[len(hashes)-1], len(hashes)-1)))
+	return hFunc.Sum()
+}
+
+// Shift lsb N Bytes to the left, with tailing 0s.
+func NBytesLeftShift(api frontend.API, input frontend.Variable, N int) frontend.Variable {
+	var lsb frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+
+	// Note: here we make sure all the calculations are within 32 bytes.
+	if N >= 32 {
+		return 0
+	}
+	for i := 0; i < 8*N; i++ {
+		lsb = api.Add(lsb, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+	multiplier = 1
+	for i := 0; i < 32-N; i++ {
+		multiplier = api.Mul(multiplier, 256)
+	}
+	return api.Mul(lsb, multiplier)
+}
+
+// Return the result of left shift by 1 byte, input size = 32 bytes (fixed)
+// Note: The MSB should be 0, or overflow occurs (the overflow behavior is not simply `mod`` by 2^256)
+// Input: {0, 42, 194, .. X other variables .., 231}
+// Output: {42, 194, , .. X other variables .., 231, 0}
+func leftShift1Byte(api frontend.API, input frontend.Variable) frontend.Variable {
+	return api.Mul(input, 256)
+}
+
+// Validate2 check the proof against the provided root commitments in a binary trie
+// ps0-3 are the proof sets. The proof sets consist of children, terminal, and path fragments along the path.
+// For example, for binary trie, there are two children (ps0, ps1), terminal (ps2), and path fragment (ps3).
+// We name them as proof sets according to the original gnark example by using a binary complete tree.
+// paths indicate the children location through the path from the leaf to the root
+func Validate2(api frontend.API, hFunc mimc.MiMC, root frontend.Variable,
+	ps0, ps1, ps2, ps3 []frontend.Variable, paths []frontend.Variable) {
+	h := hashVectors(api, hFunc, ps0[0], ps1[0], ps2[0], ps3[0])
+	for i := 1; i < len(ps0); i++ {
+		s0 := api.Select(paths[i-1], ps0[i], h)
+		s1 := api.Select(paths[i-1], h, ps1[i])
+		tmp := hashVectors(api, hFunc, s0, s1, ps2[i], ps3[i])
+		h = api.Select(api.Cmp(api.Add(ps0[i], ps1[i], ps2[i], ps3[i]), 0), tmp, h)
```

Comment:
i assume ps0[i], ps1[i], ps2[i], ps3[i] are 0 if there is no elements for that path step, thus h from the previous step is caried forward? can we add  a comment explicitly explaining this

Reply:
Thanks for pointing this out. More explanations are added in [3e590cd](https://github.com/iotaledger/trie.go/pull/5/commits/3e590cddf155314e506aabf83a80af2e21a9c96b).